meson.build

```diff
@@ -0,0 +1,284 @@
+  'silk/float/noise_shape_analysis_FLP.c',
+  'silk/float/process_gains_FLP.c',
+  'silk/float/regularize_correlations_FLP.c',
+  'silk/float/residual_energy_FLP.c',
+  'silk/float/warped_autocorrelation_FLP.c',
+  'silk/float/wrappers_FLP.c',
+  'silk/float/autocorrelation_FLP.c',
+  'silk/float/burg_modified_FLP.c',
+  'silk/float/bwexpander_FLP.c',
+  'silk/float/energy_FLP.c',
+  'silk/float/inner_product_FLP.c',
+  'silk/float/k2a_FLP.c',
+  'silk/float/LPC_inv_pred_gain_FLP.c',
+  'silk/float/pitch_analysis_core_FLP.c',
+  'silk/float/scale_copy_vector_FLP.c',
+  'silk/float/scale_vector_FLP.c',
+  'silk/float/schur_FLP.c',
+  'silk/float/sort_FLP.c'
+]
+
+silk_sources_sse4_1 = [
+  'silk/x86/NSQ_sse4_1.c',
+  'silk/x86/NSQ_del_dec_sse4_1.c',
+  'silk/x86/x86_silk_map.c',
+  'silk/x86/VAD_sse4_1.c',
+  'silk/x86/VQ_WMat_EC_sse4_1.c'
+]
+
+silk_sources += silk_sources_float
+
+opus_sources = [
+  'src/opus.c',
+  'src/opus_decoder.c',
+  'src/opus_encoder.c',
+  'src/opus_multistream.c',
+  'src/opus_multistream_encoder.c',
+  'src/opus_multistream_decoder.c',
+  'src/repacketizer.c',
+  'src/opus_projection_encoder.c',
+  'src/opus_projection_decoder.c',
+  'src/mapping_matrix.c'
+]
+
+opus_sources_float = [
+  'src/analysis.c',
+  'src/mlp.c',
+  'src/mlp_data.c'
+]
+
+opus_sources += opus_sources_float
+
+celt_sources = [
+  'celt/bands.c',
+  'celt/celt.c',
+  'celt/celt_encoder.c',
+  'celt/celt_decoder.c',
+  'celt/cwrs.c',
+  'celt/entcode.c',
+  'celt/entdec.c',
+  'celt/entenc.c',
+  'celt/kiss_fft.c',
+  'celt/laplace.c',
+  'celt/mathops.c',
+  'celt/mdct.c',
+  'celt/modes.c',
+  'celt/pitch.c',
+  'celt/celt_lpc.c',
+  'celt/quant_bands.c',
+  'celt/rate.c',
+  'celt/vq.c'
+]
+
+celt_sources_sse = [
+  'celt/x86/x86cpu.c',
+  'celt/x86/x86_celt_map.c',
+  'celt/x86/pitch_sse.c'
+]
+
+celt_sources_sse2 = [
+  'celt/x86/pitch_sse2.c',
+  'celt/x86/vq_sse2.c'
+]
+
+celt_sources_sse4_1 = [
+  'celt/x86/celt_lpc_sse4_1.c',
+  'celt/x86/pitch_sse4_1.c'
+]
+
+cc = meson.get_compiler('c')
+
+if cc.get_argument_syntax() == 'msvc'
+  sse4_1_arg = ''
+else
+  sse4_1_arg = '-msse4.1'
+endif
+
+have_sse = cc.links('''
+  #include <xmmintrin.h>
+  #include <time.h>
+  int main() {
+    __m128 mtest;
+    mtest = _mm_set1_ps((float)time(NULL));
+    mtest = _mm_mul_ps(mtest, mtest);
+    return _mm_cvtss_si32(mtest);
+  }''')
+
+have_sse2 = cc.links('''
+  #include <emmintrin.h>
+  #include <time.h>
+  int main() {
+    __m128i mtest;
+    mtest = _mm_set1_epi32((int)time(NULL));
+    mtest = _mm_mul_epu32(mtest, mtest);
+    return _mm_cvtsi128_si32(mtest);
+  }''')
+
+have_sse4_1 = cc.links('''
+  #include <smmintrin.h>
+  #include <time.h>
+  int main() {
+    __m128i mtest;
+    mtest = _mm_set1_epi32((int)time(NULL));
+    mtest = _mm_mul_epi32(mtest, mtest);
+    return _mm_cvtsi128_si32(mtest);
+  }''',
+  args: sse4_1_arg)
+
+cdata = configuration_data()
+cdata.set('OPUS_BUILD', 1)
+
+if have_sse
+  message('SSE intrinsics available')
```

Comment:
These should be options, to permit the user to override them in order to build on one machine while targeting another machine e.g. for distributing binaries intended for many end user devices.

Note: the project's configure.ac / CMakeLists.txt currently support such options.

Reply:
I've added 'feature' options for SSE, SSE2 and SSE4.1 with 'auto' as the default (which runs the `cc.links()` checks, otherwise the user's choice gets forced).

Reply:
FYI, we discussed this on IRC, and I think it would be nice to add support in meson for `cc.links(..., required: get_option('sse'))`. However I do not consider that blocker.